docs/RFCS/time_series_culling.md

```diff
@@ -0,0 +1,349 @@
+
+This algorithm will avoid scanning keys that do not need to be rolled up; this
+is desirable, as once the culling algorithm is in place and has run once, the
+majority of time series keys will *not* need to be culled.
+
+The queue will be configured to run only on the range leader for a given range
+in order to avoid duplicate work; however, this is *not* necessary for
+correctness, as demonstrated in the [Rollup Algorithm](#rollup-algorithm)
+section below.
+
+The queue will initially be set to process replicas at the same rate as the
+replica GC queue (as of this RFC, one range per 50 milliseconds).
+
+##### Package Dependency
+
+There is one particular complication to this method: *go package dependency*.
+Knowledge on how to identify and cull time series keys is contained in the `ts`
+package, but all logic for replica queues (and all current queues) lives in
+`storage`, meaning that one of three things must happen:
+
++ `storage` can depend on `ts`. This seems to be trivially possible now, but may
+be unintuitive to those trying to understand our code-base. For reference, the
+`storage` package used to depend on the `sql` package in order to record event
+logs, but this eventually became an impediment to new development and had to be
+modified.
++ The queue logic could be implemented in `ts`, and `storage` could implement
+an interface that allows it to use the `ts` code without a dependency.
++ Parts of the `ts` package could be split off into another package that can
+intuitively live below `storage`. However, this is likely to be a considerable
+portion of `ts` in order to properly implement rollups.
+
+Tenatively, we will be attempting to use the first method and have `storage`
+depend on `ts`; if it is indeed trivially possible, this will be the fastest
+method of completing this project.
+
+#### Culling low resolution data
+
+Although the volume is much lower, low-resolution data will still build
+up indefinitely unless it is culled. This data will also be culled by the same
+algorithm outlined here; however, it will not be rolled up further, but will
+simply be deleted.
+
+## Rollup algorithm
+
+The rollup algorithm is intended to be run on a single high-resolution key
+identified by the culling algorithm. The algorithm is as follows:
+
+1. Read the data in the key. Each key represents a "slab" of high resolution
+samples captured over a wall-clock hour (up to 360 samples per hour).
+2. "Downsample" all of the data in the key into a single sample; the new sample
+will have a sum, count, min and max, computed from the samples in the original
+key.
+3. Write the computed sample as a low-resolution data point into the time series
+system; this is exactly the same process as currently recorded time series,
+except it will be writing to a different key space (with a different key
+prefix).
+4. Delete the original high-resolution key.
+
+This algorithm is safe to use, even in the case where the same key is being
```

Comment:
This isn't really correct, though I don't think we need to do anything but document that incorrectness and move on.

Consider that two replicas of the queue groom a time series concurrently. In reality, I think that those could even be two different replicas. For example, what if a cullable tail of a time series covers ten Ranges and each lease holder picks up the range at the same time? Many fun interleavings are possible. Let's limit to concurrency two:

- Replica 1 and Replica 2 get processed, discover that they have the time series.
- Replica 1 issues a Scan to get the keys to be rolled up, writes the roll-up, and deletes the keys.
- Replica 2 issues a Scan to get the keys to be rolled up. Wait, there aren't any! Let me write an empty sample and delete the keys (of which there were none).

Now you've created and immediately deleted the roll-up.

It should be very rare that two Replicas of the same Range race in this way. I'm less sure about time series which span ranges, or maybe I'm operating on false assumptions on how that works in practice.

In either case, there's something to discuss and document here.